core/lib/snapshots_applier/src/lib.rs

```diff
@@ -261,6 +261,41 @@ impl SnapshotsApplierTask {
         }
     }
 
+    /// Checks whether the snapshot recovery is already completed.
+    ///
+    /// Returns `None` if no snapshot recovery information is detected in the DB.
+    /// Returns `Some(true)` if the recovery is completed.
+    /// Returns `Some(false)` if the recovery is not completed.
+    pub async fn is_recovery_completed(
+        conn: &mut Connection<'_, Core>,
+        client: &dyn SnapshotsApplierMainNodeClient,
+    ) -> anyhow::Result<Option<bool>> {
+        let Some(applied_snapshot_status) = conn
+            .snapshot_recovery_dal()
+            .get_applied_snapshot_status()
+            .await?
+        else {
+            return Ok(None);
+        };
+        // If there are unprocessed storage logs chunks, the recovery is not complete.
+        if applied_snapshot_status.storage_logs_chunks_left_to_process() != 0 {
+            return Ok(Some(false));
+        }
+        // Currently, migrating tokens is the last step of the recovery.
+        // The number of tokens is not a part of the snapshot header, so we have to re-query the main node.
+        let added_tokens = conn.tokens_dal().get_all_l2_token_addresses().await?.len();
+        let tokens_on_main_node = client
+            .fetch_tokens(applied_snapshot_status.l2_block_number)
+            .await?
+            .len();
```

Comment:
It's a bit wasteful, but I think it's OK, given that we will only reach this point _after_ we made sure that all the storage logs have been applied.
I don't feel like using a static `OnceCell` for caching here, and introducing a dedicated structure feels a bit like overkill.

Reply:
Also, this check is to be invoked if there are no blocks in the storage, so realistically we will invoke this method 1-2 times per actual snapshot recovery.